
// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

option objc_class_prefix = "HLW";

package RpcModule;

// The greeting service definition.
service Rafter {
  // 如果这个为空，则视为心跳包，可以携带信息
  rpc Appendntries (AppendntriesRequest) returns (AppendntriesReply) {}
  rpc RequestVote (RequestVoteRequest) returns (RequestVoteReply) {}
  rpc InstallSnapshot (InstallSnapshotRequest)returns (InstallSnapshotReply){}
}

message Entry{
  string cmd=1;
  int32 term=2;
  int32 index=3;
}

message StateMachine{
  int32 index=1;
  int32 term=2;
  string data=3;
}

// The request message containing the user's name.
message AppendntriesRequest {
  int32 term=1;
  int32 leaderId=2;
  int32 prevLogIndex=3;
  int32 prevLogTerm=4;
  repeated Entry entries=5;
  int32 commitIndex=6;
}

// The response message containing the greetings
message AppendntriesReply {
  int32 term=1;
  bool success=2;
  int32 matchIndex=3;
  bool installsnapshot=4;
}

message RequestVoteRequest{
  int32 term=1;
  int32 candidateId=2;
  int32 lastLogIndex=3;
  int32 lastLogTerm=4; 
}

message RequestVoteReply{
  int32 term=1;
  bool voteGranted=2; 
}

message InstallSnapshotRequest{
  int32 term=1;
  int32 leaderId=2;
  int32 lastIncludeIndex=3;//替换此索引之前的所有日志,着一个include应该是commited之前的所有日志
  int32 lastIncludeTerm=4;//lastIndluceIndex的任期
  repeated StateMachine statemachine=5;
  repeated Entry entries=6; //applied_index后面还没commit的日志
}

message InstallSnapshotReply{
  int32 term=1;
}
